### PR TITLE
Avoid lock: Call FlushStateToDisk(...) regardless of fCheckForPruning

### PR DESCRIPTION
**Pull request by practicalswift**
FlushStateToDisk(...) won't do anything besides check if we need to prune if
FLUSH_STATE_NONE is given. We avoid reading the variable fCheckForPruning
which is guarded by the mutex cs_LastBlockFile.

### CODE REVIEW

**Comment by promag** [comment]
From what I can tell it's missing only one lock (in `PruneAndFlush`) maybe fix it here too?

**Comment by practicalswift** [comment]
@promag Good point. Looks good now now?

**Comment by promag** [comment]
In that case reword PR and commit, like `Add missing cs_LastBlockFile locks`

**Comment by practicalswift** [comment]
@promag Done :-)

**Comment by practicalswift** [comment]
@promag Had to revert to initial version. Encountered timeouts when adding locking in `PruneAndFlush`. What is your suggested patch? :-)

File: src/validation.cpp [hunk 1]
```diff
@@ -3232,6 +3232,7 @@ static bool AcceptBlock(const std::shared_ptr<const CBlock>& pblock, CValidation
         return AbortNode(state, std::string("System error: ") + e.what());
     }
 
+    LOCK(cs_LastBlockFile);
```
Comment (TheBlueMatt) [inline src/validation.cpp]: It seems like we could just always call FlushStateToDisk instead? It wont do anything besides check if we need to prune if FLUSH_STATE_NONE is given.

**Comment by practicalswift** [comment]
@TheBlueMatt Even better! PR updated. Please re-review :-)

**Comment by promag** [comment]
utACK 96ff97c. Please reword PR again.

File: src/validation.cpp [hunk 2]
```diff
@@ -3232,8 +3232,7 @@ static bool AcceptBlock(const std::shared_ptr<const CBlock>& pblock, CValidation
         return AbortNode(state, std::string("System error: ") + e.what());
     }
 
-    if (fCheckForPruning)
-        FlushStateToDisk(chainparams, state, FLUSH_STATE_NONE); // we just allocated more disk space for block files
+    FlushStateToDisk(chainparams, state, FLUSH_STATE_NONE); // we just allocated more disk space for block files
```
Comment (TheBlueMatt) [inline src/validation.cpp]: Comment now needs updated.

**Comment by TheBlueMatt** [comment]
utACK 96ff97cc14b483643020df2497ab06511f5d2ae8 modulo now-incorrect comment.

**Comment by practicalswift** [comment]
@TheBlueMatt Comment removed :-)

**Comment by TheBlueMatt** [comment]
utACK 932dcd49486c6616ee5bfab9588d59fd75c4f0d4

**Comment by promag** [comment]
This PR is a bit messy now. Regarding the original intent, I think it should do:
```diff
@@ -2012,6 +2012,7 @@ void FlushStateToDisk() {
 }

 void PruneAndFlush() {
+    LOCK2(cs_main, cs_LastBlockFile);
     CValidationState state;
     fCheckForPruning = true;
     const CChainParams& chainparams = Params();
```
An alternative is to add an argument to `FlushStateToDisk`, like `bool force_check_for_pruning` to avoid touching `fCheckForPruning`.

**Comment by TheBlueMatt** [comment]
FlushStateToDisk already checks fCheckForPruning and doesn't do anything if it's not set and NONE is passed as the second argument. Despite the complication in FlushStateToDisk, I do not believe this PR has any effect except resolving the (probably not actually a) missing lock bug.

**Comment by promag** [comment]
@TheBlueMatt IMO the best approach to avoid long/recursive/shared locks is to create a snapshot of the chain and use that downstream. The only lock required is when the chain snapshot is created. A slight more advanced solution is copy-on-write, which compared to shared lock, doesn't make the  writer wait for readers to finish.

**Comment by promag** [comment]
@TheBlueMatt `PruneAndFlush` changes `fCheckForPruning` without the lock.

With:
```diff
@@ -2012,6 +2012,7 @@ void FlushStateToDisk() {
 }

 void PruneAndFlush() {
+    AssertLockHeld(cs_LastBlockFile);
     CValidationState state;
     fCheckForPruning = true;
     const CChainParams& chainparams = Params();
```
The following fails:
```
./test/functional/pruning.py
2017-11-07 01:02:25.034000 TestFramework (INFO): Initializing test directory /var/folders/1v/8_69hby54nj2k3n6fywt44x80000gn/T/testgwrg340t
Assertion failed: lock cs_LastBlockFile not held in validation.cpp:2015; locks held:
```
Note `./configure CPPFLAGS=-DDEBUG_LOCKORDER`.

**Comment by TheBlueMatt** [comment]
PruneAndFlush() would change behavior in an otherwise-simple lock-fix PR, please, no.

The other part of my context is #10279, in which I want to more clearly separate out the disk operations from validation logic - thus, I'm strongly averse to adding a disk-management lock in validation functions directly.

**Comment by practicalswift** [comment]
We have one utACK from @TheBlueMatt. Anyone else willing to review? :-)

**Comment by promag** [comment]
Testing `fCheckForPruning` is not needed in `AcceptBlock` (which BTW asserts `cs_main` is held) since the check is done in `FlushStateToDisk`.

ACK 932dcd4.

**Comment by practicalswift** [comment]
Any chance of this getting merged or should I close?

**Comment by practicalswift** [comment]
Rebased! Please re-review :-)

**Comment by practicalswift** [comment]
This one line fix is blocking my other locking annotations work since I have to re-include this fix in all other PR:s to get correct locking. Having this merged into master would help the locking annotations effort :-)

Ping @MarcoFalke, @TheBlueMatt and @laanwj :-)

**Comment by TheBlueMatt** [comment]
utACK 48c8610a9c42df6de5dead0f98dc6ced214345ba

**Comment by practicalswift** [comment]
Rebased! Please re-review :-)

**Comment by practicalswift** [comment]
Any chance of getting this merged or re-reviewed?

**Comment by sdaftuar** [comment]
utACK.  I'd prefer if we also added clearer documentation that FlushStateMode::NONE behaves in the way that we're relying on here (perhaps an extra comment in FlushStateToDisk?).

**Comment by practicalswift** [comment]
@sdaftuar Good point! Documentation added. Please re-review :-)

**Comment by sdaftuar** [comment]
Thanks, re-utACK https://github.com/bitcoin/bitcoin/pull/11617/commits/0000d8f727be937baa1499a6d15efa5a48b0ffdd

**Pull request merged**